modules/gateway/peers.go

```diff
@@ -39,6 +41,7 @@ func (g *Gateway) removePeer(peer modules.NetAddress) error {
 	}
 	delete(g.peers, peer)
 	g.save()
+	g.log.Println("INFO: removed peer", peer)
```

Comment:
there will be a log of every time an RPC returned an error, which counts as a strike. So if you want to see why a peer was booted you can grep the log for its IP. Not perfect, but certainly better than passing the error messages around or storing them somewhere.

I think context is probably the hardest problem to solve in logging. The downside of top-level logging is that you have to do a lot more of it; that is, instead of one log in `removePeer`, you need a separate log for every single function that calls `removePeer`. My intuition is that you can get most of the context you need via `runtime.Caller`, but that introduces its own problems, like "when should a log report its caller, and when shouldn't it?"